<a href="https://colab.research.google.com/github/Risky4232401021/Pemrograman-Dasar/blob/main/Pertemuan%2010%20(%20Latihan%20Cleaning%20Data%20).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

**Pertama kita masukkan tabel cutomernya terlebih dahulu**


In [3]:
customers_df =pd.read_csv("/content/drive/MyDrive/Matakuliah Pemerograman Dasar/PERTEMUAN 10 DATA/customers.csv")
customers_df.head()

customer_id customer_name             gender  age  \
0            1       fulan 1             Female   30   
1            2       fulan 2  Prefer not to say   69   
2            3       fulan 3  Prefer not to say   59   
3            4       fulan 4  Prefer not to say   67   
4            5       fulan 5  Prefer not to say   30   

                     home_address  zip_code               city  \
0  8606 Victoria TerraceSuite 560      5464      Johnstonhaven   
1      8327 Kirlin SummitApt. 461      8223    New Zacharyfort   
2       269 Gemma SummitSuite 109      5661           Aliburgh   
3       743 Bailey GroveSuite 141      1729  South Justinhaven   
4         48 Hyatt ManorSuite 375      4032     Griffithsshire   

                          state    country      status  
0            Northern Territory  Australia      Active  
1               South Australia  Australia  Non Active  
2  Australian Capital Territory  Australia  Non Active  
3                    Queensland  Australia  Non Active  
4                    Queensland  Australia  Non Active

**Kemudian kita masukkan tabel ordernya**

In [4]:
orders_df =pd.read_csv("/content/drive/MyDrive/Matakuliah Pemerograman Dasar/PERTEMUAN 10 DATA/orders.csv")
orders_df.head()

order_id  customer_id  payment  order_date delivery_date  delivery_time
0         1           64    30811  2021-08-30    2021-09-24           25.0
1         2          473    50490  2021-02-03    2021-02-13           10.0
2         3          774    46763  2021-10-08    2021-11-03           26.0
3         4          433    39782  2021-05-06    2021-05-19           13.0
4         5          441    14719  2021-03-23    2021-03-24            1.0

**Kemudian kita masukkan tabel product nya**

In [5]:
products_df =pd.read_csv("/content/drive/MyDrive/Matakuliah Pemerograman Dasar/PERTEMUAN 10 DATA/products.csv")
products_df.head()

product_id product_type  product_name size colour  price  quantity  \
0           0        Shirt  Oxford Cloth   XS    red    114        66   
1           1        Shirt  Oxford Cloth    S    red    114        53   
2           2        Shirt  Oxford Cloth    M    red    114        54   
3           3        Shirt  Oxford Cloth    L    red    114        69   
4           4        Shirt  Oxford Cloth   XL    red    114        47   

                                    description  
0  A red coloured, XS sized, Oxford Cloth Shirt  
1   A red coloured, S sized, Oxford Cloth Shirt  
2   A red coloured, M sized, Oxford Cloth Shirt  
3   A red coloured, L sized, Oxford Cloth Shirt  
4  A red coloured, XL sized, Oxford Cloth Shirt

**Lalu masukkan tabel salesnya**

In [6]:
sales_df =pd.read_csv("/content/drive/MyDrive/Matakuliah Pemerograman Dasar/PERTEMUAN 10 DATA/sales.csv")
sales_df.head()

sales_id  order_id  product_id  price_per_unit  quantity  total_price
0         0         1         218             106         2          212
1         1         1         481             118         1          118
2         2         1           2              96         3          288
3         3         1        1002             106         2          212
4         4         1         691             113         3          339

**Setelah taabelnya kita masukkan, lalu kita akan assesing data. Dimana data awal yang dianalisis adalah customers_df. Untuk mengetahui gambaran umum data tersebut, seperti jumlah kolom, tipe data, dan keberadaan nilai kosong, digunakan method info( ).**

In [7]:
customers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1001 entries, 0 to 1000
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customer_id    1001 non-null   int64 
 1   customer_name  1001 non-null   object
 2   gender         983 non-null    object
 3   age            1001 non-null   int64 
 4   home_address   1001 non-null   object
 5   zip_code       1001 non-null   int64 
 6   city           1001 non-null   object
 7   state          1001 non-null   object
 8   country        1001 non-null   object
 9   status         1001 non-null   object
dtypes: int64(3), object(7)
memory usage: 78.3+ KB


**Secara keseluruhan, tipe data pada setiap kolom sudah sesuai. Namun, perbedaan jumlah data pada kolom genderdibandingkan kolom lainnya. Ini menunjukkan adanya missing values pada kolom gender. Untuk memastikannya, dilakukan pengecekan dengan menggunakan kode berikut.**

In [8]:
customers_df.isna().sum()

customer_id       0
customer_name     0
gender           18
age               0
home_address      0
zip_code          0
city              0
state             0
country           0
status            0
dtype: int64

**Terdapat 18 data missing value pada kolom gender yang akan kita betulkan pada tahap data cleaning. Selanjutnya, perlu dilakukan pemeriksaan duplikasi data menggunakan method df.duplicated( ), dengan menjalankan kode berikut.**

In [9]:
print("Jumlah duplikasi: ", customers_df.duplicated().sum())

Jumlah duplikasi:  0


**Setelah menjalankan kode tersebut, ditemukan 6 data duplikat. Data ini akan dihapus pada tahap data cleaning. Selanjutnya, dilakukan pemeriksaan parameter statistik pada kolom numerik dalam customers_df menggunakan method describe( ). Method ini berfungsi untuk menampilkan ringkasan statistik, seperti mean dan median, dari setiap kolom numerik.**

In [10]:
customers_df.describe()

customer_id          age     zip_code
count  1001.000000  1001.000000  1001.000000
mean    500.942058    49.874126  5000.693307
std     289.013599    17.644663  2886.084454
min       1.000000    20.000000     2.000000
25%     251.000000    34.000000  2398.000000
50%     501.000000    50.000000  5079.000000
75%     751.000000    65.000000  7454.000000
max    1000.000000    80.000000  9998.000000

**Jika kita perhatikan, terdapat nilai maksimum yang tidak wajar pada kolom age. Hal ini kemungkinan disebabkan oleh adanya nilai tidak akurat pada kolom tersebut dan akan ditangani pada tahap data cleaning.**

**Selanjutnya, kita melakukan analisis pada orders_df dengan terlebih dahulu memeriksa tipe data setiap kolom menggunakan method info( ).**

In [11]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   order_id       1000 non-null   int64  
 1   customer_id    1000 non-null   int64  
 2   payment        1000 non-null   int64  
 3   order_date     1000 non-null   object 
 4   delivery_date  1000 non-null   object 
 5   delivery_time  1000 non-null   float64
dtypes: float64(1), int64(3), object(2)
memory usage: 47.0+ KB


**Bisa kita lihat bahwa seluruh data pada orders_df sudah lengkap tanpa missing value. Namun, tipe data pada kolom order_date dan delivery_date masih belum sesuai karena seharusnya menggunakan tipe datetime, bukan object.**

**Selanjutnya kita melakukan pemeriksaan duplikasi data serta ringkasan parameter statistik pada kolom numerik di orders_df.**

In [12]:
print("Jumlah duplikasi: ",orders_df.duplicated().sum())

Jumlah duplikasi:  0


In [13]:
orders_df.describe()

order_id  customer_id       payment  delivery_time
count  1000.000000  1000.000000   1000.000000    1000.000000
mean    500.500000   506.640000  33972.936000      14.078000
std     288.819436   277.115502  14451.609047       7.707225
min       1.000000     1.000000  10043.000000       1.000000
25%     250.750000   275.250000  21329.250000       8.000000
50%     500.500000   515.000000  33697.500000      14.000000
75%     750.250000   737.250000  46249.000000      21.000000
max    1000.000000  1000.000000  59910.000000      27.000000

**Berdasarkan hasil pemeriksaan, kita tidak menemukan duplikasi maupun nilai yang tidak wajar pada orders_df.**

**Selanjutnya, kita melakukan analisis pada product_df dengan memastikan kesesuaian tipe data setiap kolom menggunakan method info( ).**

In [14]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1260 entries, 0 to 1259
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   product_id    1260 non-null   int64 
 1   product_type  1260 non-null   object
 2   product_name  1260 non-null   object
 3   size          1260 non-null   object
 4   colour        1260 non-null   object
 5   price         1260 non-null   int64 
 6   quantity      1260 non-null   int64 
 7   description   1260 non-null   object
dtypes: int64(3), object(5)
memory usage: 78.9+ KB


**Dari hasil analisis, tipe data pada setiap kolom product_df sudah sesuai dan seluruh data telah lengkap tanpa adanya missing value.**

In [15]:
print("Jumlah duplikasi: ", products_df.duplicated().sum())

Jumlah duplikasi:  0


In [16]:
products_df.describe()

product_id        price     quantity
count  1260.000000  1260.000000  1260.000000
mean    629.500000   105.805556    60.150000
std     363.874979     9.704423    11.670573
min       0.000000    90.000000    40.000000
25%     314.750000    95.750000    50.000000
50%     629.500000   108.500000    60.000000
75%     944.250000   114.000000    70.000000
max    1259.000000   119.000000    80.000000

**Dari hasil pemeriksaan, terdapat 6 data duplikat pada product_df yang akan dihapus pada tahap data cleaning. Selanjutnya, kita melakukan penilaian pada sales_df. Tahap awalnya dimulai dengan memeriksa tipe data dan jumlah data menggunakan method info( ).**

In [17]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   sales_id        5000 non-null   int64
 1   order_id        5000 non-null   int64
 2   product_id      5000 non-null   int64
 3   price_per_unit  5000 non-null   int64
 4   quantity        5000 non-null   int64
 5   total_price     5000 non-null   int64
dtypes: int64(6)
memory usage: 234.5 KB


**Dari hasil percobaan, seluruh kolom pada sales_df telah menggunakan tipe data yang sesuai. Namun, jumlah data pada kolom total_price tidak sama dengan kolom lainnya, yang menandakan adanya data kosong (missing value) pada kolom tersebut.**

In [18]:
sales_df.isna().sum()

sales_id          0
order_id          0
product_id        0
price_per_unit    0
quantity          0
total_price       0
dtype: int64

**Pada kode tersebut kita menggunakan untuk menampilkan jumlah missing value pada setiap kolom di sales_df. Dari hasil yang diperoleh, diketahui terdapat 19 data hilang pada kolom total_price yang akan ditangani pada tahap data cleaning.**

**Langkah selanjutnya kita akan memeriksa duplikasi data serta ringkasan parameter statistik pada sales_df.**

In [19]:
print("Jumlah duplikasi: ", sales_df.duplicated().sum())

Jumlah duplikasi:  0


In [20]:
sales_df.describe()

sales_id     order_id   product_id  price_per_unit    quantity  \
count  5000.000000  5000.000000  5000.000000     5000.000000  5000.00000   
mean   2499.500000   503.038200   634.053200      103.501600     1.99240   
std    1443.520003   285.964418   363.255794        9.195004     0.80751   
min       0.000000     1.000000     1.000000       90.000000     1.00000   
25%    1249.750000   258.000000   323.000000       95.000000     1.00000   
50%    2499.500000   504.500000   635.000000      102.000000     2.00000   
75%    3749.250000   749.000000   951.000000      112.000000     3.00000   
max    4999.000000   999.000000  1259.000000      119.000000     3.00000   

       total_price  
count  5000.000000  
mean    206.360000  
std      86.357457  
min      90.000000  
25%     112.000000  
50%     204.000000  
75%     285.000000  
max     357.000000

**Dari hasil di atas menunjukkan bahwa tidak ada duplikasi pada sales_df**

**Berdasarkan hasil proses assessing data, terdapat tiga permasalahan pada customers_df, yaitu data duplikat, missing value, dan nilai tidak akurat. Untuk mengatasi data duplikat, digunakan method drop_duplicates( ). Berikut merupakan kode yang digunakan untuk menghapus duplikasi pada customers_df.**

In [21]:
customers_df.drop_duplicates(inplace=True)

**Setelah menjalankan kode di atas, kita cek kembali apakah masih ada data duplikat dengan menjalankan kode berikut.**

In [22]:
print("Jumlah duplikasi: ", customers_df.duplicated().sum())

Jumlah duplikasi:  0


**Jika proses penghapusan duplikasi berjalan dengan baik, maka keluaran dari kode tersebut akan menunjukkan bahwa sudah tidak ada lagi data duplikat pada customers_df.**

**Masalah selanjutnya yang perlu kita tangani adalah missing value pada kolom gender. Bisa kita lihat bahwa terdapat tiga cara untuk menangani missing value, yaitu dropping, imputation, dan interpolation. Untuk menentukan metode yang paling tepat, kita perlu melihat terlebih dahulu data yang mengandung missing value dengan melakukan filtering seperti berikut.**

In [23]:
customers_df[customers_df.gender.isna()]

customer_id customer_name gender  age                     home_address  \
38            39      fulan 39    NaN   80  7440 Cameron Estate DrSuite 628   
167          168     fulan 168    NaN   27         2781 Berge MallSuite 452   
321          322     fulan 322    NaN   30        593 Becker CircleApt. 333   
392          393     fulan 393    NaN   34          5158 Levi HillSuite 531   
441          442     fulan 442    NaN   26           5157 Feil RoadApt. 633   
719          720     fulan 720    NaN   40         31 Jordan ParadeApt. 400   
742          743     fulan 743    NaN   57   09 Christopher StreetSuite 967   
770          771     fulan 771    NaN   74     7367 Wright JunctionApt. 773   
794          795     fulan 795    NaN   49          487 Summer MewsApt. 874   
797          798     fulan 798    NaN   56           27 Aiden KnollApt. 875   
821          822     fulan 822    NaN   59        41 Jenkins KnollSuite 438   
854          855     fulan 855    NaN   55       603 O'keefe KnollSuite 782   
858          859     fulan 859    NaN   38            32 Isla GroveApt. 078   
908          909     fulan 909    NaN   62    976 Murray Station StApt. 036   
928          929     fulan 929    NaN   68           394 Lily HillSuite 153   
942          943     fulan 943    NaN   64       3117 Heller PlaceSuite 149   
946          946     fulan 946    NaN   24      8227 Nicholas HillSuite 150   
988          988     fulan 988    NaN   35   1130 Turner Estate DrSuite 925   

     zip_code                   city                         state    country  \
38       4622  North Victoriachester            Northern Territory  Australia   
167      1975         North Leoburgh             Western Australia  Australia   
321      1640             Jacobiview             Western Australia  Australia   
392      1474             Johnsburgh                    Queensland  Australia   
441      7249             Port Chloe               New South Wales  Australia   
719      1380             West Henry               South Australia  Australia   
742      6226         Lake Lukemouth             Western Australia  Australia   
770      8882               Kuhntown                      Victoria  Australia   
794      1712            East Hayden  Australian Capital Territory  Australia   
797      6531               Port Sam  Australian Capital Territory  Australia   
821      2588        Lake Andrewport               South Australia  Australia   
854      8822        Port Dylanmouth                      Tasmania  Australia   
858      7711            Rosechester               New South Wales  Australia   
908      3227               Langfort                      Tasmania  Australia   
928      2353             Beahanfurt            Northern Territory  Australia   
942       822           North Elijah               South Australia  Australia   
946       115           South Jasper                    Queensland  Australia   
988      9386              New Harry             Western Australia  Australia   

         status  
38   Non Active  
167  Non Active  
321  Non Active  
392  Non Active  
441      Active  
719  Non Active  
742  Non Active  
770      Active  
794      Active  
797  Non Active  
821  Non Active  
854  Non Active  
858  Non Active  
908      Active  
928      Active  
942  Non Active  
946      Active  
988      Active

**Kode di atas hanya akan menampilkan baris data yang memenuhi kondisi customers_df.gender.isna( ), yaitu baris data yang memiliki missing value pada kolom gender.**

**Berdasarkan hasil tersebut, terlihat bahwa data yang mengandung missing value masih memiliki banyak informasi penting. Oleh karena itu, data tersebut tidak langsung dihapus, melainkan akan ditangani menggunakan metode imputation. Gunakanlah method value_counts()untuk mengidentifikasi nilai yang dominan.**

In [24]:
customers_df.gender.value_counts()

gender
Prefer not to say    725
Male                 143
Female               115
Name: count, dtype: int64

**Berdasarkan hasil tersebut, dapat kita tahu bahwa nilai yang paling sering muncul pada kolom gender adalah “Prefer not to say”. Nilai ini kemudian digunakan untuk menggantikan missing value pada kolom tersebut. Proses penggantian dilakukan menggunakan method fillna( ), seperti pada contoh berikut.**

**Untuk memastikan proses tersebut berjalan dengan baik, kita dapat mengecek kembali keberadaan missing value**

In [25]:
customers_df.isna().sum()

customer_id       0
customer_name     0
gender           18
age               0
home_address      0
zip_code          0
city              0
state             0
country           0
status            0
dtype: int64

**Selanjutnya, kita akan mengatasi masalah inaccurate value pada kolom age. Sebagai langkah awal, kita perlu melihat baris data yang mengandung nilai tidak wajar tersebut, yaitu baris dengan nilai age maksimum. Untuk itu, digunakan teknik filtering seperti kode ini.**

In [26]:
customers_df[customers_df.age == customers_df.age.max()]

customer_id customer_name             gender  age  \
34            35      fulan 35               Male   80   
38            39      fulan 39                NaN   80   
142          143     fulan 143  Prefer not to say   80   
154          155     fulan 155               Male   80   
170          171     fulan 171  Prefer not to say   80   
174          175     fulan 175  Prefer not to say   80   
181          182     fulan 182               Male   80   
235          236     fulan 236  Prefer not to say   80   
423          424     fulan 424  Prefer not to say   80   
437          438     fulan 438  Prefer not to say   80   
494          495     fulan 495               Male   80   
523          524     fulan 524  Prefer not to say   80   
596          597     fulan 597  Prefer not to say   80   
701          702     fulan 702  Prefer not to say   80   
731          732     fulan 732  Prefer not to say   80   

                          home_address  zip_code                   city  \
34   3168 Bartoletti CrescentSuite 878       593             Port Lucas   
38     7440 Cameron Estate DrSuite 628      4622  North Victoriachester   
142        87 Hahn Station StSuite 943      8254               New Ella   
154           85 Charles MallSuite 424      7841            Port Sophia   
170              16 Boyer WaySuite 018      6226          South Gabriel   
174             57 David MallSuite 596      3129                Haneton   
181           32 Thomas CrestSuite 753      2147             Archiefurt   
235            64 Phillips RunApt. 722      2752            East Sienna   
423            03 Grant CrestSuite 228      5743      Lake Charliemouth   
437          569 Alyssa IslandApt. 766      6893             Hunterland   
494          1798 Schultz MewsApt. 704      9409           Lake Caitlin   
523            0358 Ashley WayApt. 506      3345     Port Sebastianland   
596          78 O'hara CircuitApt. 096      1966       Lake Michaelbury   
701              61 Bosco WaySuite 775       685        Thompsonborough   
731    905 Stephanie BoulevardApt. 217      8114          Port Aaronton   

                  state    country      status  
34           Queensland  Australia      Active  
38   Northern Territory  Australia  Non Active  
142            Victoria  Australia  Non Active  
154     New South Wales  Australia      Active  
170   Western Australia  Australia      Active  
174     South Australia  Australia  Non Active  
181          Queensland  Australia  Non Active  
235     New South Wales  Australia      Active  
423   Western Australia  Australia  Non Active  
437   Western Australia  Australia  Non Active  
494     New South Wales  Australia  Non Active  
523          Queensland  Australia      Active  
596     South Australia  Australia      Active  
701   Western Australia  Australia  Non Active  
731          Queensland  Australia      Active

**Kode di atas akan menampilkan baris data dengan nilai age paling besar.Dari data tersebut, dapat diasumsikan bahwa nilai age yang tidak biasa terjadi akibat human error, yaitu kelebihan memasukkan angka nol. Oleh karena itu, nilai tersebut akan diganti menjadi 70 menggunakan method replace( ), seperti pada kode berikut.**

In [27]:
customers_df.age.replace(customers_df.age.max(), 70, inplace=True)

/tmp/ipython-input-4147959128.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  customers_df.age.replace(customers_df.age.max(), 70, inplace=True)


In [28]:
customers_df[customers_df.age == customers_df.age.max()]

customer_id customer_name             gender  age  \
17            18      fulan 18               Male   79   
27            28      fulan 28  Prefer not to say   79   
47            48      fulan 48               Male   79   
68            69      fulan 69  Prefer not to say   79   
73            74      fulan 74  Prefer not to say   79   
151          152     fulan 152             Female   79   
209          210     fulan 210  Prefer not to say   79   
211          212     fulan 212  Prefer not to say   79   
219          220     fulan 220  Prefer not to say   79   
303          304     fulan 304  Prefer not to say   79   
403          404     fulan 404  Prefer not to say   79   
443          444     fulan 444  Prefer not to say   79   
559          560     fulan 560  Prefer not to say   79   
641          642     fulan 642  Prefer not to say   79   
677          678     fulan 678  Prefer not to say   79   
680          681     fulan 681  Prefer not to say   79   
698          699     fulan 699  Prefer not to say   79   
743          744     fulan 744               Male   79   
754          755     fulan 755             Female   79   
805          806     fulan 806               Male   79   

                     home_address  zip_code               city  \
17        13 Bailey ManorApt. 848      7171         Walterland   
27       360 Heller DriveApt. 052      5820          West Noah   
47         78 Nguyen MallApt. 116      4761     Heathcotemouth   
68         473 Leo CourtSuite 784      8820         East Sofia   
73   3263 Parisian PlaceSuite 811      7803      West Jesseton   
151       03 Glover TrackApt. 881      7845      Christianport   
209    476 Mueller CrestSuite 674      5663  West Claudiaburgh   
211    385 Jayden View RdApt. 459      4481         Port Logan   
219         566 Kiara RunApt. 417      1993   West Zacharybury   
303      5306 Hickle MewsApt. 405      7906        Feeneymouth   
403       29 Bruen TrailSuite 288      9569      West Ellaport   
443     361 Davies CrestSuite 392      6837     Port Andrewton   
559   4674 Victoria DriveApt. 830      6122    Port Poppymouth   
641      303 O'keefe PassApt. 658       192     North Benjamin   
677    609 Rohan TerraceSuite 767      1948          East Alex   
680      096 Gibson ManorApt. 195      1323     West Alyssaton   
698       4959 James HillApt. 613      2027           Westfurt   
743      440 Ernser GroveApt. 463      9828      East Zoeshire   
754      04 Dare TerraceSuite 650      9976         West Aaron   
805        865 Ivy TrailSuite 324      1265     New Joshuaview   

                            state    country      status  
17                New South Wales  Australia      Active  
27                     Queensland  Australia      Active  
47                New South Wales  Australia  Non Active  
68                       Victoria  Australia      Active  
73             Northern Territory  Australia  Non Active  
151  Australian Capital Territory  Australia  Non Active  
209               New South Wales  Australia  Non Active  
211             Western Australia  Australia      Active  
219             Western Australia  Australia  Non Active  
303                    Queensland  Australia  Non Active  
403  Australian Capital Territory  Australia      Active  
443             Western Australia  Australia  Non Active  
559                    Queensland  Australia  Non Active  
641                      Tasmania  Australia  Non Active  
677               New South Wales  Australia      Active  
680                      Victoria  Australia      Active  
698             Western Australia  Australia  Non Active  
743            Northern Territory  Australia  Non Active  
754               New South Wales  Australia      Active  
805                      Tasmania  Australia  Non Active

**Ternyata, masih ditemukan invalid value lain pada kolom age. Kemungkinan penyebabnya sama seperti sebelumnya, yaitu human error akibat kelebihan memasukkan angka nol. Untuk mengatasinya, nilai tersebut akan diganti menjadi 50.**

In [29]:
customers_df.age.replace(customers_df.age.max(), 50, inplace=True)

/tmp/ipython-input-606824494.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  customers_df.age.replace(customers_df.age.max(), 50, inplace=True)


In [30]:
customers_df[customers_df.age == customers_df.age.max()]

customer_id customer_name             gender  age  \
11            12      fulan 12  Prefer not to say   78   
62            63      fulan 63  Prefer not to say   78   
87            88      fulan 88               Male   78   
97            98      fulan 98             Female   78   
172          173     fulan 173  Prefer not to say   78   
217          218     fulan 218  Prefer not to say   78   
223          224     fulan 224  Prefer not to say   78   
250          251     fulan 251             Female   78   
258          259     fulan 259  Prefer not to say   78   
336          337     fulan 337  Prefer not to say   78   
402          403     fulan 403  Prefer not to say   78   
404          405     fulan 405               Male   78   
468          469     fulan 469             Female   78   
499          500     fulan 500  Prefer not to say   78   
538          539     fulan 539               Male   78   
612          613     fulan 613  Prefer not to say   78   
679          680     fulan 680  Prefer not to say   78   
683          684     fulan 684  Prefer not to say   78   

                       home_address  zip_code                city  \
11           66 Kayla MewsSuite 951      6334         Taylorburgh   
62        013 Edward ParadeApt. 499      5140            Kossfort   
87     331 Leah Station StSuite 445      3111           Port Liam   
97        772 Chloe MeadowSuite 666      2945     East Siennaview   
172          5133 Mann MallApt. 757      5630       Port Maxville   
217           88 Walsh WaySuite 074      7479          Lake Lucas   
223         20 Xavier RidgeApt. 805      7744          Hansenberg   
250    786 Mackenzie StreetApt. 660      8520          Wilsonside   
258      205 Bailey IslandSuite 339       752           Kingmouth   
336          88 Olson GroveApt. 120      2631          Cooperbury   
402        902 Rees CircleSuite 205      1038           Port Abby   
404        41 Noah TerraceSuite 130      5174        North Nathan   
468    783 Olivia CrescentSuite 877      5011       Andrewborough   
499          74 Aaron HillSuite 794      1819  Port Haydenchester   
538         6481 Isaac LoopApt. 355      4362          Port Blake   
612  2525 Leannon JunctionSuite 996      6475           Stellaton   
679      5530 Ashton GroveSuite 128      7905          Taylorfurt   
683     2079 Phoebe AvenueSuite 629      8834          Laurenview   

                            state    country      status  
11                South Australia  Australia      Active  
62                       Victoria  Australia  Non Active  
87                South Australia  Australia      Active  
97                     Queensland  Australia  Non Active  
172               South Australia  Australia  Non Active  
217                    Queensland  Australia      Active  
223               New South Wales  Australia      Active  
250  Australian Capital Territory  Australia      Active  
258               New South Wales  Australia  Non Active  
336             Western Australia  Australia      Active  
402             Western Australia  Australia  Non Active  
404               New South Wales  Australia  Non Active  
468               New South Wales  Australia      Active  
499             Western Australia  Australia      Active  
538                    Queensland  Australia      Active  
612               New South Wales  Australia  Non Active  
679               New South Wales  Australia      Active  
683             Western Australia  Australia      Active

**Pastikan lagi bahwa tidak ada data yang inaccurate pada kolom age**

In [31]:
customers_df.describe()

customer_id          age     zip_code
count  1001.000000  1001.000000  1001.000000
mean    500.942058    49.144855  5000.693307
std     289.013599    16.920343  2886.084454
min       1.000000    20.000000     2.000000
25%     251.000000    34.000000  2398.000000
50%     501.000000    50.000000  5079.000000
75%     751.000000    63.000000  7454.000000
max    1000.000000    78.000000  9998.000000

**Berdasarkan hasil tersebut, bisa kita lihat bahwa nilai maksimum pada kolom age sudah cukup masuk akal. Selain itu, jika diperhatikan, nilai mean dan standard deviation juga ikut berubah setelah penanganan inaccurate value dilakukan.**

**Setelah menyelesaikan seluruh permasalahan pada customers_df, tahap berikutnya adalah menangani data pada orders_df. Berdasarkan proses assessing data sebelumnya, ditemukan kesalahan tipe data pada kolom order_date dan delivery_date.**

**Untuk memperbaiki hal tersebut, tipe data kedua kolom tersebut akan diubah menjadi datetime. Proses ini dilakukan menggunakan fungsi to_datetime() yang tersedia pada library pandas.**

In [32]:
datetime_columns = ["order_date", "delivery_date"]

for column in datetime_columns:
  orders_df[column] = pd.to_datetime(orders_df[column])

**Kode di atas kita gunakan untuk mengubah tipe data pada kolom order_date dan delivery_date menjadi datetime. Untuk memastikan proses tersebut berjalan sesuai dengan yang diharapkan, kita dapat mengecek kembali tipe data kolom tersebut menggunakan method info( ).**

In [33]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   order_id       1000 non-null   int64         
 1   customer_id    1000 non-null   int64         
 2   payment        1000 non-null   int64         
 3   order_date     1000 non-null   datetime64[ns]
 4   delivery_date  1000 non-null   datetime64[ns]
 5   delivery_time  1000 non-null   float64       
dtypes: datetime64[ns](2), float64(1), int64(3)
memory usage: 47.0 KB


**Jika semua langkah berhasil dijalankan, maka tipe data pada kedua kolom tersebut kini sudah berubah menjadi datetime.**

**Data selanjutnya yang akan kita bersihkan adalah product_df. Berdasarkan hasil assessing data sebelumnya, diketahui terdapat 6 data duplikat pada product_df. Untuk mengatasinya, data yang terduplikasi tersebut akan dihapus menggunakan method drop_duplicates( ), seperti pada kode ini.**

In [34]:
products_df.drop_duplicates(inplace=True)

**Kode tersebut digunakan untuk menghilangkan semua data yang terduplikasi. Untuk memastikan hasilnya sudah sesuai, kita bisa melakukan pengecekan kembali dengan menjalankan kode berikut.**

In [35]:
print("Jumlah duplikasi: ", products_df.duplicated().sum())

Jumlah duplikasi:  0


**Jika proses penghapusan data duplikat berhasil, maka keluaran dari kode tersebut akan menunjukkan “Jumlah duplikasi: 0”, seperti hasil outputnya.**

**Data selanjutnya yang akan kita bersihkan adalah sales_df. Dari hasil penilaian data sebelumnya, diketahui terdapat 19 missing value pada kolom total_price. Untuk menentukan cara penanganan yang tepat, kita perlu melihat terlebih dahulu baris data yang mengandung missing value tersebut.**

In [36]:
sales_df[sales_df.total_price.isna()]

Empty DataFrame
Columns: [sales_id, order_id, product_id, price_per_unit, quantity, total_price]
Index: []

**Kode di atas digunakan untuk menampilkan seluruh baris data yang memiliki nilai kosong (missing value) pada kolom total_price, seperti yang terlihat pada data di atas. Dari hasil tampilan data tersebut, dapat diketahui bahwa nilai total_price diperoleh dari hasil perkalian antara price_per_unit dan quantity.**

**Kode ini dapat dimanfaatkan untuk mengatasi missing value pada kolom total_price. Berikut ini merupakan contoh penerapan kode yang digunakan untuk melakukan proses tersebut.**

In [37]:
sales_df["total_price"] = sales_df["price_per_unit"] * sales_df["quantity"]

**Kode di atas kita gunakan untuk mengatasi semua missing value serta memastikan bahwa nilai pada kolom total_price sudah benar. Untuk memastikan hasilnya, kita bisa mengecek kembali jumlah missing value pada sales_df menggunakan kode berikut.**

In [38]:
sales_df.isna().sum()

sales_id          0
order_id          0
product_id        0
price_per_unit    0
quantity          0
total_price       0
dtype: int64

**Jika proses sebelumnya berjalan dengan baik, maka hasil yang diperoleh akan sesuai seperti pada data di atas. Tahapan tersebut merupakan bagian dari proses data wrangling yang dilakukan untuk menyiapkan data sebelum tahap analisis.**

**Setelah memahami teori terkait proses EDA, selanjutnya kita akan menerapkannya pada sebuah proyek analisis data sederhana.**

In [39]:
customers_df.describe(include="all")

customer_id customer_name             gender          age  \
count   1001.000000          1001                983  1001.000000   
unique          NaN          1000                  3          NaN   
top             NaN     fulan 943  Prefer not to say          NaN   
freq            NaN             2                725          NaN   
mean     500.942058           NaN                NaN    49.144855   
std      289.013599           NaN                NaN    16.920343   
min        1.000000           NaN                NaN    20.000000   
25%      251.000000           NaN                NaN    34.000000   
50%      501.000000           NaN                NaN    50.000000   
75%      751.000000           NaN                NaN    63.000000   
max     1000.000000           NaN                NaN    78.000000   

                      home_address     zip_code        city            state  \
count                         1001  1001.000000        1001             1001   
unique                        1000          NaN         961                8   
top     3117 Heller PlaceSuite 149          NaN  East Aidan  South Australia   
freq                             2          NaN           3              140   
mean                           NaN  5000.693307         NaN              NaN   
std                            NaN  2886.084454         NaN              NaN   
min                            NaN     2.000000         NaN              NaN   
25%                            NaN  2398.000000         NaN              NaN   
50%                            NaN  5079.000000         NaN              NaN   
75%                            NaN  7454.000000         NaN              NaN   
max                            NaN  9998.000000         NaN              NaN   

          country  status  
count        1001    1001  
unique          1       2  
top     Australia  Active  
freq         1001     617  
mean          NaN     NaN  
std           NaN     NaN  
min           NaN     NaN  
25%           NaN     NaN  
50%           NaN     NaN  
75%           NaN     NaN  
max           NaN     NaN

**Berdasarkan rangkuman parameter statistik di atas, kita akan memperoleh informasi jumlah pelanggan sebanyak 1001 orang yang berumur di kisaran antara 20 hingga 80 tahun dengan rata-rata umur sebesar 49.87 tahun dengan standar deviasinya sebesar 17.64 tahun. Informasi ini tentunya dapat memberikan gambaran bahwa segmentasi pelanggan yang kita miliki cukup luas mulai dari remaja hingga lanjut usia.**

**Sekarang coba kita lihat demografi pelanggan berdasarkan jenis kelamin (gender). Untuk melakukan ini, kita akan menggunakan method groupby() yang diikuti dengan method agg(). Berikut contoh kode untuk melakukannya.**

In [40]:
customers_df.groupby(by="gender").agg({
    "customer_id": "nunique",
    "age": ["max", "min", "mean", "std"]
})

customer_id age                          
                      nunique max min       mean        std
gender                                                     
Female                    115  78  20  48.643478  16.162655
Male                      143  78  20  50.139860  17.452794
Prefer not to say         725  78  20  49.051034  16.965643

In [41]:
customers_df.groupby(by="city").customer_id.nunique().sort_values(ascending=False)

city
East Sophia     3
New Ava         3
East Aidan      3
Lake Sienna     2
Jademouth       2
               ..
Justinport      1
Kaiberg         1
Kaifort         1
Kaihaven        1
Johnsonhaven    1
Name: customer_id, Length: 961, dtype: int64

In [42]:
customers_df.groupby(by="state").customer_id.nunique().sort_values(ascending=False)

state
South Australia                 139
Queensland                      134
New South Wales                 132
Northern Territory              125
Western Australia               124
Australian Capital Territory    121
Victoria                        121
Tasmania                        104
Name: customer_id, dtype: int64

**Operasi yang akan kita lakukan ialah menghitung jumlah detik dari delivery_time menggunakan method total_seconds(). selanjutnya diubah ke dalam satuan hari (dibagi 86400) dan diambil bilangan bulatnya**

In [43]:
delivery_time = orders_df["delivery_date"] - orders_df["order_date"]
delivery_time = delivery_time.apply(lambda x: x.total_seconds())
orders_df["delivery_time"] = round(delivery_time/86400)

**Kode di atas akan menghasilkan kolom baru pada DataFrame orders_df.**

In [44]:
orders_df.describe(include="all")

order_id  customer_id       payment                  order_date  \
count  1000.000000  1000.000000   1000.000000                        1000   
mean    500.500000   506.640000  33972.936000  2021-05-27 18:38:52.800000   
min       1.000000     1.000000  10043.000000         2021-01-01 00:00:00   
25%     250.750000   275.250000  21329.250000         2021-03-13 18:00:00   
50%     500.500000   515.000000  33697.500000         2021-05-27 12:00:00   
75%     750.250000   737.250000  46249.000000         2021-08-12 06:00:00   
max    1000.000000  1000.000000  59910.000000         2021-10-24 00:00:00   
std     288.819436   277.115502  14451.609047                         NaN   

             delivery_date  delivery_time  
count                 1000    1000.000000  
mean   2021-06-10 20:31:12      14.078000  
min    2021-01-03 00:00:00       1.000000  
25%    2021-03-28 00:00:00       8.000000  
50%    2021-06-11 12:00:00      14.000000  
75%    2021-08-24 06:00:00      21.000000  
max    2021-11-20 00:00:00      27.000000  
std                    NaN       7.707225

**Berdasarkan hasil tersebut, dapat diketahui bahwa rata-rata waktu pengiriman adalah 14 hari, dengan waktu pengiriman terlama mencapai 27 hari dan tercepat sebesar 1**

**Selanjutnya ekspoloring data kita dapat menambahkan kolom baru bernama status pada data customers_df. Kolom ini berisi nilai “Active” untuk pelanggan yang setidaknya pernah melakukan satu kali order, sedangkan “Non Active” diberikan kepada pelanggan yang belum pernah melakukan order sama sekali.**

In [45]:
customer_id_in_orders_df =  orders_df.customer_id.tolist()
customers_df["status"] = customers_df["customer_id"].apply(lambda x: "Active" if x in customer_id_in_orders_df else "Non Active")
customers_df.sample(5)

customer_id customer_name             gender  age  \
928          929     fulan 929                NaN   68   
564          565     fulan 565  Prefer not to say   73   
851          852     fulan 852  Prefer not to say   67   
253          254     fulan 254               Male   44   
969          969     fulan 969  Prefer not to say   77   

                     home_address  zip_code          city               state  \
928        394 Lily HillSuite 153      2353    Beahanfurt  Northern Territory   
564      27 Kiara MeadowSuite 662      4126  Hamiltonport          Queensland   
851        079 Chase MallApt. 902      8460  O'connertown            Tasmania   
253  4966 Lilly Estate DrApt. 570      8170    Beiermouth          Queensland   
969     08 Howe BoulevardApt. 286      9561   West Stella     South Australia   

       country  status  
928  Australia  Active  
564  Australia  Active  
851  Australia  Active  
253  Australia  Active  
969  Australia  Active

**Kode di atas akan menghasilkan lima sample data, Bisa di ubah parameter 5 nya sesuai dengan keinginan untuk melihat data yg lebih banyak**

In [46]:
customers_df.groupby(by="status").customer_id.count()

status
Active        617
Non Active    384
Name: customer_id, dtype: int64

**Selanjutnya agar kita bisa mendapatkan informasi yang lebih lengkap dari kedua data tersebut, kita perlu menggabungkannya melalui proses join atau merge. Berikut ini merupakan contoh kode yang digunakan untuk melakukan merge pada data orders_df dan customers_df.**

In [47]:
orders_customers_df = pd.merge(
    left=orders_df,
    right=customers_df,
    how="left",
    left_on="customer_id",
    right_on="customer_id"
)
orders_customers_df.head()

order_id  customer_id  payment order_date delivery_date  delivery_time  \
0         1           64    30811 2021-08-30    2021-09-24           25.0   
1         2          473    50490 2021-02-03    2021-02-13           10.0   
2         3          774    46763 2021-10-08    2021-11-03           26.0   
3         4          433    39782 2021-05-06    2021-05-19           13.0   
4         5          441    14719 2021-03-23    2021-03-24            1.0   

  customer_name             gender  age                   home_address  \
0      fulan 64  Prefer not to say   75      4927 Alice MeadowApt. 960   
1     fulan 473               Male   61  531 Schmitt BoulevardApt. 010   
2     fulan 774  Prefer not to say   34       2096 Wilson MewsApt. 714   
3     fulan 433  Prefer not to say   46        5777 Mayer PassApt. 881   
4     fulan 441  Prefer not to say   53   33 Richards JunctionApt. 478   

   zip_code                city              state    country  status  
0      7787      Sanfordborough    South Australia  Australia  Active  
1      1744             Annaton    South Australia  Australia  Active  
2      8590  West Jonathanshire           Tasmania  Australia  Active  
3      9728    West Michaelport           Tasmania  Australia  Active  
4      7650     South Rileyview  Western Australia  Australia  Active

**Jumlah order berdasarkan kota Kita bisa membuat pivot table untuk memperoleh informasi terkait jumlah order berdasarkan kota dengan kode ini.**

In [48]:
orders_customers_df.groupby(by="city").order_id.nunique().sort_values(ascending=False).reset_index().head(10)

city  order_id
0        Jordanside         6
1           New Ava         6
2  Port Hannahburgh         5
3         Lake Rose         5
4        O'keefeton         5
5          East Max         5
6          Rubyfort         5
7          West Kai         5
8         Sarahstad         4
9        Russelstad         4

**Berdasarkan hasil tersebut, diketahui bahwa kota Jordanside dan New Ava merupakan dua kota yang memiliki jumlah order terbanyak.**

**Selanjutnya, kita akan melihat jumlah order berdasarkan state. Untuk memperoleh informasi tersebut, kita dapat menggunakan pivot table. Gunakan kode di bawah ini untuk menerapkannya.**

In [49]:
orders_customers_df.groupby(by="state").order_id.nunique().sort_values(ascending=False)

state
South Australia                 148
Queensland                      139
Western Australia               130
New South Wales                 129
Victoria                        118
Australian Capital Territory    118
Tasmania                        112
Northern Territory              106
Name: order_id, dtype: int64

**Kode tersebut akan menghasilkan pivot table seperti di bawah ini. Berdasarkan pivot table tersebut, diketahui bahwa South Australia merupakan state yang paling banyak melakukan order.**

**Informasi selanjutnya yang dapat kita analisis adalah jumlah order berdasarkan gender. Untuk melihat hal tersebut, kita dapat membuat sebuah pivot table dengan menggunakan kode berikut.**

In [50]:
orders_customers_df.groupby(by="gender").order_id.nunique().sort_values(ascending=False)

gender
Prefer not to say    716
Female               139
Male                 136
Name: order_id, dtype: int64

**Berdasarkan hasil tersebut, dapat dilihat bahwa kebanyakan order dilakukan oleh pelanggan yang berjenis kelamin prefer not to say. Hal ini tentunya sejalan dengan jumlah pelanggan yang kebanyakan dari kelompok gender tersebut.**

**Selanjutnya, kita akan mengeksplorasi jumlah order berdasarkan kelompok usia. Untuk itu, perlu dibuat kolom baru bernama age_group yang digunakan untuk mengelompokkan pelanggan ke dalam tiga kategori, yaitu youth, adults, dan seniors. Setelah itu, kita dapat membuat pivot table berdasarkan kelompok usia tersebut menggunakan contoh kode di bawah ini.**

In [51]:
orders_customers_df["age_group"] = orders_customers_df.age.apply(lambda x: "Youth" if x <= 24 else ("Seniors" if x > 64 else "Adults"))

In [52]:
orders_customers_df.groupby(by="age_group").order_id.nunique().sort_values(ascending=False)

age_group
Adults     691
Seniors    216
Youth       93
Name: order_id, dtype: int64

**Berdasarkan hasil tersebut diketahui bahwa pelanggan yang paling banyak melakukan order berasal dari kelompok usia Adults.**

**Selanjutnya, kita akan mengeksplorasi data product_df dan sales_df. Sebagai langkah awal, kita dapat melihat ringkasan parameter statistik dari kedua data tersebut dengan menggunakan method describe( ).**

In [53]:
products_df.describe(include="all")

product_id product_type product_name  size colour        price  \
count   1260.000000         1260         1260  1260   1260  1260.000000   
unique          NaN            3           35     8      7          NaN   
top             NaN        Shirt        Denim    XS    red          NaN   
freq            NaN          420           70   252    180          NaN   
mean     629.500000          NaN          NaN   NaN    NaN   105.805556   
std      363.874979          NaN          NaN   NaN    NaN     9.704423   
min        0.000000          NaN          NaN   NaN    NaN    90.000000   
25%      314.750000          NaN          NaN   NaN    NaN    95.750000   
50%      629.500000          NaN          NaN   NaN    NaN   108.500000   
75%      944.250000          NaN          NaN   NaN    NaN   114.000000   
max     1259.000000          NaN          NaN   NaN    NaN   119.000000   

           quantity                                        description  
count   1260.000000                                               1260  
unique          NaN                                               1260  
top             NaN  A violet coloured, XL sized, Tracksuit Bottoms...  
freq            NaN                                                  1  
mean      60.150000                                                NaN  
std       11.670573                                                NaN  
min       40.000000                                                NaN  
25%       50.000000                                                NaN  
50%       60.000000                                                NaN  
75%       70.000000                                                NaN  
max       80.000000                                                NaN

In [54]:
sales_df.describe(include="all")

sales_id     order_id   product_id  price_per_unit    quantity  \
count  5000.000000  5000.000000  5000.000000     5000.000000  5000.00000   
mean   2499.500000   503.038200   634.053200      103.501600     1.99240   
std    1443.520003   285.964418   363.255794        9.195004     0.80751   
min       0.000000     1.000000     1.000000       90.000000     1.00000   
25%    1249.750000   258.000000   323.000000       95.000000     1.00000   
50%    2499.500000   504.500000   635.000000      102.000000     2.00000   
75%    3749.250000   749.000000   951.000000      112.000000     3.00000   
max    4999.000000   999.000000  1259.000000      119.000000     3.00000   

       total_price  
count  5000.000000  
mean    206.360000  
std      86.357457  
min      90.000000  
25%     112.000000  
50%     204.000000  
75%     285.000000  
max     357.000000

**Dari hasil tersebut, dapat dilihat bahwa harga produk yang dijual berada pada kisaran 90 hingga 119 dolar. Selain itu, diperoleh informasi lain yang cukup menarik, yaitu dalam satu transaksi pelanggan maksimal membeli tiga item untuk satu jenis produk dengan total harga mencapai 357 dolar.**

**Jika kita ingin mengetahui produk dengan harga tertinggi dan terendah, silakan jalankan kode ini untuk melihat hasilnya.**

In [55]:
products_df.sort_values(by="price", ascending=False)

product_id product_type product_name size  colour  price  quantity  \
665          665       Jacket        Parka   XS     red    119        65   
666          666       Jacket        Parka    S     red    119        66   
667          667       Jacket        Parka    M     red    119        56   
668          668       Jacket        Parka    L     red    119        73   
669          669       Jacket        Parka   XL     red    119        64   
...          ...          ...          ...  ...     ...    ...       ...   
1216        1216     Trousers      Pleated    S  indigo     90        41   
1223        1223     Trousers      Pleated    L  violet     90        48   
1222        1222     Trousers      Pleated    M  violet     90        45   
1221        1221     Trousers      Pleated    S  violet     90        50   
1220        1220     Trousers      Pleated   XS  violet     90        58   

                                        description  
665          A red coloured, XS sized, Parka Jacket  
666           A red coloured, S sized, Parka Jacket  
667           A red coloured, M sized, Parka Jacket  
668           A red coloured, L sized, Parka Jacket  
669          A red coloured, XL sized, Parka Jacket  
...                                             ...  
1216   A indigo coloured, S sized, Pleated Trousers  
1223   A violet coloured, L sized, Pleated Trousers  
1222   A violet coloured, M sized, Pleated Trousers  
1221   A violet coloured, S sized, Pleated Trousers  
1220  A violet coloured, XS sized, Pleated Trousers  

[1260 rows x 8 columns]

**Selanjutnya, kita dapat memanfaatkan pivot table untuk memperoleh informasi produk berdasarkan tipe dan nama produk. inilah untuk kodenya.**

In [56]:
products_df.groupby(by="product_type").agg({
    "product_id": "nunique",
    "quantity": "sum",
    "price":  ["min", "max"]
})

product_id quantity price     
                nunique      sum   min  max
product_type                               
Jacket              420    25387    90  119
Shirt               420    25263    92  119
Trousers            420    25139    90  119

In [57]:
products_df.groupby(by="product_name").agg({
    "product_id": "nunique",
    "quantity": "sum",
    "price": ["min", "max"]
})

product_id quantity price     
                     nunique      sum   min  max
product_name                                    
Bomber                    35     2083    90   90
Camp Collared             35     2071   112  112
Cardigan                  35     2032   118  118
Cargo Pants               35     2146   106  106
Casual Slim Fit           35     2086   113  113
Chambray                  35     2020   105  105
Chinos                    35     2101   100  100
Coach                     35     2158   115  115
Cords                     35     2260   113  113
Cropped                   35     2085    99   99
Cuban Collar              35     2017    93   93
Denim                     70     4204    92  115
Drawstring                35     2042   104  104
Dress                     35     2125   119  119
Flannel                   35     2135    96   96
Henley                    35     2051    92   92
High-Waisted              35     2198    98   98
Joggers                   35     2107    94   94
Leather                   35     2276   113  113
Linen                     35     2138   116  116
Mandarin Collar           35     2203   108  108
Oxford Cloth              35     2071   114  114
Parka                     35     2201   119  119
Peacoat                   35     2067   102  102
Pleated                   35     2030    90   90
Polo                      35     2196   117  117
Puffer                    35     1946   110  110
Pullover                  35     2184   114  114
Relaxed Leg               35     2002    95   95
Shearling                 35     2169    95   95
Slim-Fit                  35     2155   119  119
Tracksuit Bottoms         35     2038    91   91
Trench Coat               35     2132   112  112
Windbreaker               35     2085   109  109
Wool                      35     1975   111  111

**Pivot table di atas memberikan gambaran mengenai produk-produk yang dijual oleh Fashion Collection. Sebagai calon praktisi data yang andal, tentu kita penasaran produk mana yang paling laris. Untuk menjawab hal tersebut, kita perlu menggabungkan (merge) tabel product_df dan sales_df menggunakan berikut.**

In [58]:
sales_products_df = pd.merge(
    left=sales_df,
    right=products_df,
    how="left",
    left_on="product_id",
    right_on="product_id"
)
sales_products_df.head()

sales_id  order_id  product_id  price_per_unit  quantity_x  total_price  \
0         0         1         218             106           2          212   
1         1         1         481             118           1          118   
2         2         1           2              96           3          288   
3         3         1        1002             106           2          212   
4         4         1         691             113           3          339   

  product_type  product_name size  colour  price  quantity_y  \
0        Shirt      Chambray    L  orange    105          44   
1       Jacket        Puffer    S  indigo    110          62   
2        Shirt  Oxford Cloth    M     red    114          54   
3     Trousers          Wool    M    blue    111          52   
4       Jacket         Parka    S  indigo    119          53   

                                   description  
0   A orange coloured, L sized, Chambray Shirt  
1    A indigo coloured, S sized, Puffer Jacket  
2  A red coloured, M sized, Oxford Cloth Shirt  
3      A blue coloured, M sized, Wool Trousers  
4     A indigo coloured, S sized, Parka Jacket

**Kemudian kita akan melihat informasi penjualan produk berdasarkan tipe dengan membuat pivot table berdasarkan product type menggunakan kode berikut.**

In [59]:
sales_products_df.groupby(by="product_type").agg({
    "sales_id": "nunique",
    "quantity_x": "sum",
    "total_price": "sum"
})

sales_id  quantity_x  total_price
product_type                                   
Jacket            1676        3343       357026
Shirt             1641        3259       333600
Trousers          1683        3360       341174

In [60]:
sales_products_df.groupby(by="product_name").agg({
    "sales_id": "nunique",
    "quantity_x": "sum",
    "total_price": "sum"
}).sort_values(by="total_price", ascending=False)

sales_id  quantity_x  total_price
product_name                                        
Denim                   273         527        52399
Casual Slim Fit         154         306        36414
Trench Coat             146         299        35581
Shearling               150         302        35334
Puffer                  140         298        35164
Flannel                 141         281        33158
Cropped                 135         284        32660
Pleated                 147         308        32340
Joggers                 164         334        31062
Chambray                141         290        30740
Parka                   134         269        30397
Bomber                  150         297        29700
Chinos                  135         260        29380
Windbreaker             143         287        29274
Cardigan                134         260        29120
Peacoat                 132         266        28994
Tracksuit Bottoms       130         257        28784
Slim-Fit                154         288        27936
Oxford Cloth            146         289        27744
Drawstring              147         288        27648
High-Waisted            140         278        27522
Camp Collared           139         279        27342
Cargo Pants             134         267        27234
Relaxed Leg             146         301        27090
Cuban Collar            122         254        26670
Henley                  146         281        26414
Wool                    125         249        26394
Leather                 136         263        25774
Mandarin Collar         124         236        25724
Dress                   127         243        25272
Pullover                143         276        25116
Polo                    131         264        25080
Coach                   134         265        24645
Linen                   131         270        24570
Cords                   126         246        23124

**Selanjutnya kita akan membuat sebuah DataFrame baru bernama all_df untuk menampung seluruh informasi dari keempat tabel yang dimiliki.**

In [61]:
all_df = pd.merge(
    left=sales_products_df,
    right=orders_customers_df,
    how="left",
    left_on="order_id",
    right_on="order_id"
)
all_df.head()

sales_id  order_id  product_id  price_per_unit  quantity_x  total_price  \
0         0         1         218             106           2          212   
1         1         1         481             118           1          118   
2         2         1           2              96           3          288   
3         3         1        1002             106           2          212   
4         4         1         691             113           3          339   

  product_type  product_name size  colour  ...  customer_name  \
0        Shirt      Chambray    L  orange  ...       fulan 64   
1       Jacket        Puffer    S  indigo  ...       fulan 64   
2        Shirt  Oxford Cloth    M     red  ...       fulan 64   
3     Trousers          Wool    M    blue  ...       fulan 64   
4       Jacket         Parka    S  indigo  ...       fulan 64   

              gender age               home_address  zip_code            city  \
0  Prefer not to say  75  4927 Alice MeadowApt. 960      7787  Sanfordborough   
1  Prefer not to say  75  4927 Alice MeadowApt. 960      7787  Sanfordborough   
2  Prefer not to say  75  4927 Alice MeadowApt. 960      7787  Sanfordborough   
3  Prefer not to say  75  4927 Alice MeadowApt. 960      7787  Sanfordborough   
4  Prefer not to say  75  4927 Alice MeadowApt. 960      7787  Sanfordborough   

             state    country  status age_group  
0  South Australia  Australia  Active   Seniors  
1  South Australia  Australia  Active   Seniors  
2  South Australia  Australia  Active   Seniors  
3  South Australia  Australia  Active   Seniors  
4  South Australia  Australia  Active   Seniors  

[5 rows x 28 columns]

**Bisa kita lihat Proses merge di atas akan menghasilkan DataFrame seperti hasil di atas**

**Kemudian kita akan melihat preferensi pembelian berdasarkan state pelanggan dan tipe produknya**

In [62]:
all_df.groupby(by=["state", "product_type"]).agg({
    "quantity_x": "sum",
    "total_price": "sum"
})

quantity_x  total_price
state                        product_type                         
Australian Capital Territory Jacket               406        43204
                             Shirt                396        40448
                             Trousers             454        46790
New South Wales              Jacket               451        47998
                             Shirt                431        43980
                             Trousers             392        39766
Northern Territory           Jacket               365        38991
                             Shirt                336        33865
                             Trousers             384        38998
Queensland                   Jacket               499        53511
                             Shirt                417        42506
                             Trousers             453        46045
South Australia              Jacket               461        49090
                             Shirt                509        52685
                             Trousers             455        46041
Tasmania                     Jacket               412        44370
                             Shirt                387        39668
                             Trousers             379        38057
Victoria                     Jacket               359        38203
                             Shirt                390        40206
                             Trousers             414        41948
Western Australia            Jacket               390        41659
                             Shirt                393        40242
                             Trousers             429        43529

**Dan terakhir kita dapat melakukan analisis yang sama untuk mengetahui preferensi tipe produk pelanggan berdasarkan gender dan kelompok usia.**

In [63]:
all_df.groupby(by=["gender", "product_type"]).agg({
    "quantity_x": "sum",
    "total_price": "sum"
})

quantity_x  total_price
gender            product_type                         
Female            Jacket               481        50963
                  Shirt                445        45146
                  Trousers             481        49295
Male              Jacket               480        50978
                  Shirt                410        41939
                  Trousers             453        46071
Prefer not to say Jacket              2349       251486
                  Shirt               2386       244765
                  Trousers            2400       243092

In [64]:
all_df.groupby(by=["age_group", "product_type"]).agg({
    "quantity_x": "sum",
    "total_price": "sum"
})

quantity_x  total_price
age_group product_type                         
Adults    Jacket              2332       249475
          Shirt               2255       230907
          Trousers            2317       235880
Seniors   Jacket               737        78539
          Shirt                703        71851
          Trousers             721        72862
Youth     Jacket               274        29012
          Shirt                301        30842
          Trousers             322        32432